src/components/implementation/no_interface/cbboot/quarantine.c

```diff
@@ -218,45 +226,44 @@ quarantine_fork(spdid_t spdid, spdid_t source)
 				struct cbid_caddr new_cbm;
 				new_cbm.caddr = cbuf_alloc_ext(left, &new_cbm.cbid, CBUF_EXACTSZ);
 				printd("Avoid sharing, allocated new cbuf %d @ %x\n", new_cbm.cbid, (unsigned long)new_cbm.caddr);
-				if (!new_cbm.caddr) return 0;
-				memcpy(new_cbm.caddr, cbm.caddr, left);
-				cbm.caddr = new_cbm.caddr;
+				if (!new_cbm.caddr) { printc("error 7\n"); goto error; }
+				memcpy(new_cbm.caddr, cbm.caddr, left);										// only memcpy
+				cbm.caddr = new_cbm.caddr;											// use cbuf instead? Copy on write???
 				cbm.cbid = new_cbm.cbid;
 			}
 			assert(cbm.caddr);
 			cbid = cbm.cbid;
 			new_sect_cbufs[j] = cbm;
 
-			if (d_addr != (cbuf_map_at(cos_spd_id(), cbid, d_spd, d_addr | flags))) return 0;
+			if (d_addr != (cbuf_map_at(cos_spd_id(), cbid, dest, d_addr | flags))) { printc("error 8\n"); goto error;}
 			if (sect->flags & COBJ_SECT_CINFO) {
 				/* fixup cinfo page */
 				struct cos_component_information *ci = cbm.caddr;
-				ci->cos_this_spd_id = d_spd;
-				__boot_spd_set_symbs(h, d_spd, ci);
-				__boot_deps_save_hp(d_spd, ci->cos_heap_ptr);
+				ci->cos_this_spd_id = dest;
+				__boot_spd_set_symbs(src_hdr, dest, ci);
+				__boot_deps_save_hp(dest, ci->cos_heap_ptr);
 			}
 			prev_map += left - PAGE_SIZE;
 			d_addr += left;
 		}
-
 	}
 
 	/* FIXME: set fault handlers and re-write caps */
-	printd("Activating %d\n", d_spd);
-	if (cos_spd_cntl(COS_SPD_ACTIVATE, d_spd, h->ncap, 0)) BUG();
-	printd("Setting capabilities for %d\n", d_spd);
-	if (__boot_spd_caps(h, d_spd)) BUG();
+	printd("Activating %d\n", dest);
+	if (cos_spd_cntl(COS_SPD_ACTIVATE, dest, src_hdr->ncap, 0)) BUG();
+	printd("Setting capabilities for %d\n", dest);
+	if (__boot_spd_caps(src_hdr, dest)) BUG();
 
 	/* Increment send-side fork count in source's caps, and
-	 * copy fork count into d_spd's caps. */
-	for (j = 0; j < h->ncap; j++) {
+	 * copy fork count into dest's caps. */
+	for (j = 0; j < src_hdr->ncap; j++) {
 		int cnt;
-		cap = cobj_cap_get(h, j);
-		if (cobj_cap_undef(cap)) break;
+		cap = cobj_cap_get(src_hdr, j);
+		if (cobj_cap_undef(cap)) break;									// what does this do?
```

Comment:
Some of these questions should be really trivial to answer by just looking up the function.